yarn-project/p2p/src/services/tx_collection/fast_tx_collection.ts

```diff
@@ -0,0 +1,298 @@
+  ) {}
+
+  public async stop() {
+    this.requests.forEach(request => request.promise.reject(new AbortError(`Stopped collection service`)));
+    await Promise.resolve();
+  }
+
+  public getFastCollectionRequests() {
+    return this.requests;
+  }
+
+  public async collectFastFor(
+    input: FastCollectionRequestInput,
+    txHashes: TxHash[] | string[],
+    opts: { deadline: Date; pinnedPeer?: PeerId },
+  ) {
+    const timeout = opts.deadline.getTime() - this.dateProvider.now();
+    if (timeout <= 0) {
+      this.log.warn(`Deadline for fast tx collection is in the past (${timeout}ms)`, {
+        deadline: opts.deadline.getTime(),
+        now: this.dateProvider.now(),
+      });
+      return [];
+    }
+
+    const blockInfo: BlockInfo =
+      input.type === 'proposal' ? input.blockProposal.toBlockInfo() : input.block.toBlockInfo();
+
+    // This promise use used to await for the collection to finish during the main collectFast method.
+    // It gets resolved in `foundTxs` when all txs have been collected, or rejected if the request is aborted or hits the deadline.
+    const promise = promiseWithResolvers<void>();
+    setTimeout(() => promise.reject(new TimeoutError(`Timed out while collecting txs`)), timeout);
+
+    const request: FastCollectionRequest = {
+      ...input,
+      blockInfo,
+      promise,
+      foundTxs: new Map<string, TxWithHash>(),
+      missingTxHashes: new Set(txHashes.map(t => t.toString())),
+      deadline: opts.deadline,
+    };
+
+    const [duration] = await elapsed(() => this.collectFast(request, { ...opts }));
+
+    this.log.verbose(
+      `Collected ${request.foundTxs.size} txs out of ${txHashes.length} for ${input.type} at slot ${blockInfo.slotNumber}`,
+      {
+        ...blockInfo,
+        duration,
+        requestType: input.type,
+        missingTxs: [...request.missingTxHashes],
+      },
+    );
+    return [...request.foundTxs.values()];
+  }
+
+  private async collectFast(
+    request: FastCollectionRequest,
+    opts: { proposal?: BlockProposal; deadline: Date; pinnedPeer?: PeerId },
+  ) {
+    this.requests.add(request);
+    const { blockInfo } = request;
+
+    this.log.debug(
+      `Starting fast collection of ${request.missingTxHashes.size} txs for ${request.type} at slot ${blockInfo.slotNumber}`,
+      { ...blockInfo, requestType: request.type, deadline: opts.deadline },
+    );
+
+    try {
+      // Start blasting all nodes for the txs. We give them a little time to respond before we start reqresp.
+      // And keep an eye on the request promise to ensure we don't wait longer than the deadline or return as soon
+      // as we have collected all txs, whatever the source.
+      const nodeCollectionPromise = this.collectFastFromNodes(request, opts);
+      const waitBeforeReqResp = sleep(this.config.txCollectionFastNodesTimeoutBeforeReqRespMs);
+      await Promise.race([nodeCollectionPromise, request.promise.promise, waitBeforeReqResp]);
```

Comment:
I'm not sure what do we get by not requesting both over RPC and Req/Resp immediately and concurrently?